### PR TITLE
Api

### PR DESCRIPTION
**Pull request by VladimirHaninSIB**
This is a draft !

### CODE REVIEW

File: epinorm/dataFetcher.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import pandas as pd
+from polyglot.text import Text
+from transliterate import translit
+
+def getCountryNamesStandardised():
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: In python, the naming convention is to use `snake_case` for function and variable names.

In this case, you would name your function `get_country_names_standardized`, or maybe `get_standardized_country_names`.

File: epinorm/dataFetcher.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    THis retrieves the country names in the file "countries", as they are in the file
+    """
+
+    df = pd.read_csv("epinorm/data/countries.csv")
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: I would consider storing the path to the file in some sort of global variable, so its location is set in one central place.

File: epinorm/dataFetcher.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    names = df["name"]
+    return set(names)
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: Here there is no need for an intermediate `names` variable I think.
```suggestion
    return set(df["name"])
```

Or you could even do 
```py
    return set(pd.read_csv("epinorm/data/countries.csv")["name"])
```

File: epinorm/dataFetcher.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    countryToCode["United States"] = "US"
+    countryToCode["US"] = "US"
+    countryToCode["Venezuela"] = "VE"
+    countryToCode["Vietnam"] = "VN"
+
+
+    # now map country code to admin levels
+    countryCodeToAdminLevels = {}
+
+    df = pd.read_csv("epinorm/data/administrative_units.tsv", sep="\t")
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: Same comment here as above regarding storing the path in a global variable.  
Also, we could use the same format (`.csv`) for all files.

File: epinorm/dataFetcher.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    return set(names)
+
+def getAdminLevelsOfCountryTable():
+
+    # * create a mapping from country name to its code
+    countryToCode = {}
+
+    df = pd.read_csv("epinorm/data/countries.csv")
+    for _, row in df.iterrows():
+        countryToCode[row["name"]] = row["alpha_2"]
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: I think there is a method to directly convert a pandas DataFrame to a `dict`, so you don't need a loop here:

* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_dict.html

File: epinorm/dataFetcher.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    # there are exceptions
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: How about modifying this directly in the source file `epinorm/data/countries.csv` ?
Or is it important to keep this file as is ?

File: epinorm/dataFetcher.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    """
+
+    df = pd.read_csv("epinorm/data/countries.csv")
+    names = df["name"]
+    return set(names)
+
+def getAdminLevelsOfCountryTable():
+
+    # * create a mapping from country name to its code
+    countryToCode = {}
```
Comment (robinengler) [inline epinorm/dataFetcher.py]: Same comment about following python `snake_case` convention for variable names.
I would rename this to `country_codes`, or `codes_by_country`.

**Comment by VladimirHaninSIB** [comment]
refactoring and optimisations were implemented

File: epinorm/geo.py [hunk 8]
```diff
@@ -21,7 +22,7 @@
 USER_AGENT = "MOOD Geocoder"
 DEFAULT_ZOOM_LEVEL = 10
 DEFAULT_RESULT_LIMIT = 1
-REMOTE_REQUEST_DELAY = 2
+REMOTE_REQUEST_DELAY = 2 # in second
```
Comment (robinengler) [inline epinorm/geo.py]:
```suggestion
REMOTE_REQUEST_DELAY = 2 # in seconds
```

File: epinorm/geo.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 
 from calendar import timegm
 from time import gmtime, sleep
+from time import gmtime, sleep, time
```
Comment (robinengler) [inline epinorm/geo.py]:
```suggestion
from time import gmtime, sleep, time
```

File: epinorm/geo.py [hunk 10]
```diff
@@ -30,9 +31,17 @@
...
+    
+    TIME_LAST_REQUEST = time()
 
     def fetch(self, url, params=None):
         """Fetch data from the web."""
+
+        elaped_time_since_last_request = time() - self.TIME_LAST_REQUEST
+        if elaped_time_since_last_request < REMOTE_REQUEST_DELAY: # must wait before making new request
+            sleep(REMOTE_REQUEST_DELAY - elaped_time_since_last_request)
+        self.TIME_LAST_REQUEST = time()
```
Comment (robinengler) [inline epinorm/geo.py]: Just a typo:

```suggestion
        elapsed_time_since_last_request = time() - self.TIME_LAST_REQUEST
        if elapsed_time_since_last_request < REMOTE_REQUEST_DELAY: # must wait before making new request
            sleep(REMOTE_REQUEST_DELAY - elapsed_time_since_last_request)
        self.TIME_LAST_REQUEST = time()
```

File: epinorm/geo.py [hunk 11]
```diff
@@ -30,9 +31,17 @@
 
 
 class Geocoder:
+    
+    TIME_LAST_REQUEST = time()
```
Comment (robinengler) [inline epinorm/geo.py]: As far as I know, the python convention is to use UPPER_CASE_NAMES only for naming global constants. Here we are dealing with a regular class attribute, so it should be in lower case I think.

```suggestion
    time_last_request = time()
```

File: epinorm/norm.py [hunk 12]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+        It returns a dictionnary that maps a country name (from countries.csv) to a list of its administrative levels.
+        administrative levels are encoded as dictionaries with properties:
+        {
+            name,
+            admin_level,
+            osmId
+        }
+
+        it also tranliterates information from the administrative_units.tsv file.
+        """
```
Comment (robinengler) [inline epinorm/norm.py]: A few typos. Also, we (or at least I) try to adhere to the 80 chars line length limit.

```suggestion
        """
        Retrieves the information in "countries.csv" and
        "administrative_units.tsv", and returns a dictionary that maps a
        country name (from "countries.csv") to a list of its administrative
        levels.

        Administrative levels are encoded as dictionaries with properties:
        {
            name,
            admin_level,
            osmId
        }

        Also transliterates information from the administrative_units.tsv file.
        """
```

File: epinorm/norm.py [hunk 13]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+
+        # * create a mapping from country name to its two letter code
+        country_to_code = {}
+        df = pd.read_csv(COUNTRIES_DATA)
+        for _, row in df.iterrows():
+            country_to_code[row["name"]] = row["alpha_2"]
```
Comment (robinengler) [inline epinorm/norm.py]: Here we can use `.to_dict()` to avoid a `for` loop.

```suggestion
        # Create a mapping from country name to its two letter code.
        country_to_code = pd.read_csv(COUNTRIES_DATA, index_col="name")["alpha_2"].to_dict()
```

File: epinorm/norm.py [hunk 14]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+
+        # add the exceptions
+        country_to_code.update(COUNTRIES_EXCEPTIONS) 
+
+
```
Comment (robinengler) [inline epinorm/norm.py]:
```suggestion

```

File: epinorm/norm.py [hunk 15]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+        # * now map country code to its admin levels
+        code_to_admin = {}
+        df = pd.read_csv(ADMIN_LEVELS_DATA, sep="\t")
+        for _, row in df.iterrows():
```
Comment (robinengler) [inline epinorm/norm.py]: The `read_table()` method of pandas uses tabs as its default separator.  
And we can also skip creating the intermediate `df` object.

```suggestion
        for _, row in pd.read_table(ADMIN_LEVELS_DATA).iterrows():
```

File: epinorm/norm.py [hunk 16]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+                    transliteration = translit(row["endonym"], reversed=True)
+
+                    # if the original language was russian, then it might have used the character ь which
+                    # doesn't get transliterate properly (it gets into an apostrophe)
+                    if "ь" in row["endonym"] and "'" in transliteration:
+                        transliteration = transliteration.replace("'", "")
+
+                    entry = {"name":transliteration, "admin_level":row["admin_level"], "osmId":row["osm_id"]}
+                    code_to_admin[row["iso3166_1_code"]].append(entry)
+                except:
```
Comment (robinengler) [inline epinorm/norm.py]: Usually we want to `except` a specific error type(s) that we expect, but not generically catch all exceptions.

**Review by robinengler — COMMENTED** [review]
Not done yet, but I want to already submit this before shutting down my machine.

File: epinorm/geo.py (hunk 10 above)
Comment (robinengler) [inline epinorm/geo.py]: As discussed earlier, we would like to revert this change to the simpler original version.

File: epinorm/norm.py [hunk 17]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
             .dropna()
             .unique()
         )
+    
+    def _admin_levels_table(self):
```
Comment (robinengler) [inline epinorm/norm.py]: I'm not sure this needs to be a method of the class?  (because it never uses `self`).  
It could just be a function in the module.

File: epinorm/norm.py [hunk 18]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+        # * now map country code to its admin levels
+        code_to_admin = {}
+        df = pd.read_csv(ADMIN_LEVELS_DATA, sep="\t")
+        for _, row in df.iterrows():
+
+                if row["iso3166_1_code"] not in code_to_admin:
+                    code_to_admin[row["iso3166_1_code"]] = []
+
+                # a row sometimes doesn't contain endonym
+                if type(row["endonym"]) is not float:
```
Comment (robinengler) [inline epinorm/norm.py]: Use `isinstance` instead of comparing types:

```py
if isinstance(row["endonym"], str):
```

File: epinorm/norm.py [hunk 19]
```diff
@@ -122,6 +129,86 @@ def get_feature_ids(self):
...
+
+                    # some words are okay to be replaced, they are modifiers
+                    replacements = {
+                    "District" : "Region",
+                    "Region" : "District",
+                    }
+                    for first, second in replacements.items():
+                        if first in row["exonym"]:
+                            entry = {"name":row["exonym"].replace(first, second), "admin_level":row["admin_level"], "osmId":row["osm_id"]}
+                            code_to_admin[row["iso3166_1_code"]].append(entry)
```
Comment (robinengler) [inline epinorm/norm.py]: I would suggest variables names that are more explicit than "first" and "second",
maybe something like:

```suggestion
                    for original, synonym in replacements.items():
                        if original in row["exonym"]:
                            entry = {"name":row["exonym"].replace(original, synonym), "admin_level":row["admin_level"], "osmId":row["osm_id"]}
                            code_to_admin[row["iso3166_1_code"]].append(entry)
```

File: epinorm/norm.py (hunk 15 above)
Comment (robinengler) [inline epinorm/norm.py]: As discussed, this loop could be rewritten using individual functions and `.apply()`.

To get rid of "non-string" (i.e. empty) values, you can use `.dropna()` (I tested on the "endonym" column and at least for that column it appears to work).

```py

def get_endonym(row):
    return {"name": row["endonym"], "admin_level":row["admin_level"], "osmId": row["osm_id"]}

def get_transliteration(row):
    ...

def get_exonym(row):
    ...

code_to_admin = {
    country_code:(
        list(df.loc[df["iso3166_1_code"] == country_code, ].apply(get_endonym, axis=1)) + 
        list(df.apply(get_transliteration, axis=1)) +
        list(df.apply(get_exonym, axis=1))
        )
    for country_code in df["iso3166_1_code"].unique()
}

code_to_admin = {}
for country_code in df["iso3166_1_code"].unique():
    country_filter = df["iso3166_1_code"] == country_code
    
    code_to_admin[country_filter] = (
        list(df.loc[country_filter, ].dropna().apply(get_endonym, axis=1)) + 
        list(df.loc[country_filter, ].dropna().apply(get_transliteration, axis=1)) +
        list(df.loc[country_filter, ].dropna().apply(get_exonym, axis=1))
    )
        



```

File: epinorm/norm.py [hunk 20]
```diff
@@ -282,24 +386,65 @@ def _get_location_from_strain(self, record):
...
-        country = str(coalesce(location["country"], ""))
-        area = str(coalesce(location["area"], ""))
-        places = []
-        for place in area.split(","):
+
+
+        # format country to make comparison easier
+        location["country"] = location["country"].strip().lower()
+
+        newPlaces = [] # combine areas and extrated location together in this list
```
Comment (robinengler) [inline epinorm/norm.py]: Use `snake_case` :snake: 

```suggestion
        new_places = [] # combine areas and extracted location together in this list
```

File: epinorm/norm.py [hunk 21]
```diff
@@ -266,12 +353,29 @@ def __init__(self, data_file):
...
-            location["area"] = places[1].strip()
+        """
+
+        location = {"country": None, "areas":[]}
+
+        if ":" in record["original_location"]:
+            objects = record["original_location"].split(":", 1)
+
+            location["country"] = objects[0]
+            location["areas"] = objects[1].split(",") # even if there isn't a comma you get what we need
```
Comment (robinengler) [inline epinorm/norm.py]: I would suggest using variables names more explicit than "objects":

```suggestion
            country, area = record["original_location"].split(":", 1)
            location["country"] = country
            location["areas"] =area.split(",") # even if there isn't a comma you get what we need
```

File: epinorm/norm.py [hunk 22]
```diff
@@ -282,24 +386,65 @@ def _get_location_from_strain(self, record):
...
+                newPlaces.append(place)
+
+
+        # format extracted field
+        if record["extracted_location"]:
+            place = record["extracted_location"].replace("_", " ")
+            place = re.sub(r"\(.+\)", "", place)
+            place = re.sub(r"[^\w\s-]+", "", place)
+            place = place.strip()
+            place = place.lower() # for easier comparison later
```
Comment (robinengler) [inline epinorm/norm.py]: Some of these lines can be combined I think.
Maybe also the 2 regular expressions.

```suggestion
            place = record["extracted_location"].replace("_", " ").strip().lower()
            place = re.sub(r"\(.+\)", "", place)
            place = re.sub(r"[^\w\s-]+", "", place)
```

File: epinorm/norm.py [hunk 23]
```diff
@@ -282,24 +386,65 @@ def _get_location_from_strain(self, record):
...
+        }
+        newPlacesAgain = []
+        for token in newPlaces:
+
+            if token in exceptionMappings :
+                newPlacesAgain.append(exceptionMappings[token])
+            else:
+                newPlacesAgain.append(token)
+            
+        location["areas"] = newPlacesAgain
```
Comment (robinengler) [inline epinorm/norm.py]: Using list comprehensions, you should be able to do something like:

```suggestion
        location["areas"] = [
            exceptionMappings[token] if token in exceptionMappings else token for token in newPlaces
        ]
```

File: epinorm/norm.py [hunk 24]
```diff
@@ -1,40 +1,138 @@
...
     "locality_osm_id",
-    "original_record_source",
-    "original_record_id",
+    "admin_level_1",
+    "admin_level_1_osm_id",
+
+    # "original_record_source",
+    # "original_record_id",
     "original_record_location_description",
 ]
```
Comment (orryn) [inline epinorm/norm.py]: Restore the original output columns and column order.

File: epinorm/norm.py [hunk 25]
```diff
@@ -1,40 +1,138 @@
...
-
+import itertools
 from csv import QUOTE_NONE
-from epinorm.config import REF_DATA_DIR
+from transliterate import translit
+from transliterate.exceptions import LanguageDetectionError
+
 from epinorm.geo import NominatimGeocoder
-from epinorm.utils import coalesce
+from epinorm.config import MIN_ADMIN_EXCEPTIONS, REF_DATA_DIR, COUNTRIES_DATA, COUNTRIES_EXCEPTIONS, ADMIN_LEVELS_DATA
```
Comment (orryn) [inline epinorm/norm.py]: Use a multi-line import statement:

```suggestion
from epinorm.config import (
    MIN_ADMIN_EXCEPTIONS,
    REF_DATA_DIR,
    COUNTRIES_DATA,
    COUNTRIES_EXCEPTIONS,
    ADMIN_LEVELS_DATA,
)
```

File: epinorm/norm.py [hunk 26]
```diff
@@ -1,40 +1,138 @@
...
-    "original_record_id",
+    "admin_level_1",
+    "admin_level_1_osm_id",
+
+    # "original_record_source",
+    # "original_record_id",
     "original_record_location_description",
 ]
 
+def get_translitaration_endonym(row):
```
Comment (orryn) [inline epinorm/norm.py]: Correct the typo in the function name and update all calls to the function. I would also recommend renaming the function to:

```suggestion
def get_transliterated_endonym(row):
```

File: epinorm/norm.py [hunk 27]
```diff
@@ -1,40 +1,138 @@
...
+                        "osm_id": row["osm_id"]}
+                new_entries.append(entry)
+
+                new_entries += get_translitaration_endonym(row)
+
+            # some words are okay to be replaced, they are modifiers
+            replacements = {
+            "District" : "Region",
+            "Region" : "District",
+            }
```
Comment (orryn) [inline epinorm/norm.py]: The insertion of "region" and "district" pairs can lead to wrong results. It is possible for a region and a district to share the same name, because the two units are at different administrative levels. The idea behind the addition of synonyms is to accommodate naming variations for the same administrative units. For example, there are records in GenBank referring to "Omsk region", which corresponds to "Omsk Oblast" in OpenStreetMap. By providing a variant in which "oblast" is replaced with "region", we can correctly resolve the locations in those records.

```suggestion
            replacements = {
            "Oblast" : "Region",
            "Krai" : "Region",
            }
```

File: epinorm/norm.py [hunk 28]
```diff
@@ -282,24 +397,63 @@ def _get_location_from_strain(self, record):
...
+
+        # format extracted field
+        if record["extracted_location"]:
+
+            cleaned_place = clean_token(record["extracted_location"])
+
+            if cleaned_place != "" and cleaned_place != location["country"] and \
+               cleaned_place not in new_places:
+                new_places.append(cleaned_place)
+
```
Comment (orryn) [inline epinorm/norm.py]: To avoid code duplication, combine location["areas"] and record["extracted_location"] in a single array and apply the processing logic once.

File: epinorm/norm.py [hunk 29]
```diff
@@ -320,27 +474,170 @@ def _add_missing_columns(self):
...
 
+    def _search_tokens_diff_order(self, areas, second):
+        """
+        This method uses the geocoder to make a query of areas (a list of tokens) and second (a string).
+        we vary the order of the tokens in areas as they could be wrong. We also remove them if we still don't have a result
+        it tried all permutations up to the point where it doesn't include tokens from areas anymore (only "second")
+        """
+
+        for l in range(len(areas), -1, -1):
+            for option in itertools.permutations(areas, l):
```
Comment (orryn) [inline epinorm/norm.py]: Ambiguous variable name (E741). Use "i" for iteration counter, instead of "l".

```suggestion
        for i in range(len(areas), -1, -1):
            for option in itertools.permutations(areas, i):
```

File: epinorm/norm.py [hunk 30]
```diff
@@ -320,27 +474,170 @@ def _add_missing_columns(self):
         self._data["longitude"] = None
         self._data["original_record_id"] = None
 
+    def _search_tokens_diff_order(self, areas, second):
```
Comment (orryn) [inline epinorm/norm.py]: To improve the code readability, rename the argument "second" to "appended_text" and update all argument references.

File: epinorm/norm.py [hunk 31]
```diff
@@ -320,27 +474,170 @@ def _add_missing_columns(self):
...
+                locality = self._geocoder.get_feature(
+                    "search", {"query": query}, term=query, term_type="query"
+                )
+                if locality:
+                    return locality
+
+        return None
+
+        
+    def _find_full_locality(self, areas, countryName):
```
Comment (orryn) [inline epinorm/norm.py]: Rename the argument "countryName" to "country_name" and update all references to it.

File: epinorm/norm.py [hunk 32]
```diff
@@ -320,27 +474,170 @@ def _add_missing_columns(self):
...
+            if len(areas) == 0 and len(admin_levels_found) == 0:
+                continue
+
+            # we are here if we haven't identified the highest admin boundary, or we might still have
+            # information on the city that wasn't matched with admin_boundaries
+
+            sorted(admin_levels_found, key=lambda x: x["admin_level"], reverse=True)
+            adminLevelsMatched = list(map(lambda x: x["name"], admin_levels_found))
+            admin_level = "" if admin_level_1s[i] is None else admin_level_1s[i]
+            second = ", ".join(adminLevelsMatched) + ", " + admin_level + ", " + country.get("name")
```
Comment (orryn) [inline epinorm/norm.py]: Use snake_case for variable names.

```suggestion
            matched_admin_levels = list(map(lambda x: x["name"], admin_levels_found))
            admin_level = "" if admin_level_1s[i] is None else admin_level_1s[i]
            second = ", ".join(matched_admin_levels) + ", " + admin_level + ", " + country.get("name")
```

**Comment by VladimirHaninSIB** [comment]
ready to merge 👍
